File: frontend/src/services/accounting/reconciliationService.js
import api from '@/api/axios';

const RECONCILIATION_BASE_URL = '/api/v1/reconciliations';

export default {
  /**
   * Get a list of reconciliations with optional filters and pagination
   * @param {Object} params - Query parameters (page, limit, status, accountId, search, etc.)
   * @returns {Promise<Array>} - List of reconciliations
   */
  async getReconciliations(params = {}) {
    try {
      const response = await api.get(RECONCILIATION_BASE_URL, { params });
      return response.data;
    } catch (error) {
      console.error('Error fetching reconciliations:', error);
      throw error;
    }
  },

  /**
   * Get a single reconciliation by ID
   * @param {string|number} id - Reconciliation ID
   * @returns {Promise<Object>} - Reconciliation details
   */
  async getReconciliation(id) {
    try {
      const response = await api.get(`${RECONCILIATION_BASE_URL}/${id}`);
      return response.data;
    } catch (error) {
      console.error(`Error fetching reconciliation ${id}:`, error);
      throw error;
    }
  },

  /**
   * Create a new reconciliation
   * @param {Object} data - Reconciliation data
   * @returns {Promise<Object>} - Created reconciliation
   */
  async createReconciliation(data) {
    try {
      const response = await api.post(RECONCILIATION_BASE_URL, data);
      return response.data;
    } catch (error) {
      console.error('Error creating reconciliation:', error);
      throw error;
    }
  },

  /**
   * Update an existing reconciliation
   * @param {string|number} id - Reconciliation ID
   * @param {Object} data - Updated reconciliation data
   * @returns {Promise<Object>} - Updated reconciliation
   */
  async updateReconciliation(id, data) {
    try {
      const response = await api.put(`${RECONCILIATION_BASE_URL}/${id}`, data);
      return response.data;
    } catch (error) {
      console.error(`Error updating reconciliation ${id}:`, error);
      throw error;
    }
  },

  /**
   * Delete a reconciliation
   * @param {string|number} id - Reconciliation ID
   * @returns {Promise<void>}
   */
  async deleteReconciliation(id) {
    try {
      await api.delete(`${RECONCILIATION_BASE_URL}/${id}`);
    } catch (error) {
      console.error(`Error deleting reconciliation ${id}:`, error);
      throw error;
    }
  },

  /**
   * Get reconciliation items for a specific reconciliation
   * @param {string|number} reconciliationId - Reconciliation ID
   * @param {Object} params - Query parameters (page, limit, status, etc.)
   * @returns {Promise<Array>} - List of reconciliation items
   */
  async getReconciliationItems(reconciliationId, params = {}) {
    try {
      const response = await api.get(
        `${RECONCILIATION_BASE_URL}/${reconciliationId}/items`,
        { params }
      );
      return response.data;
    } catch (error) {
      console.error(`Error fetching reconciliation ${reconciliationId} items:`, error);
      throw error;
    }
  },

  /**
   * Add an item to a reconciliation
   * @param {string|number} reconciliationId - Reconciliation ID
   * @param {Object} item - Item data
   * @returns {Promise<Object>} - Created item
   */
  async addReconciliationItem(reconciliationId, item) {
    try {
      const response = await api.post(
        `${RECONCILIATION_BASE_URL}/${reconciliationId}/items`,
        item
      );
      return response.data;
    } catch (error) {
      console.error(`Error adding item to reconciliation ${reconciliationId}:`, error);
      throw error;
    }
  },

  /**
   * Update a reconciliation item
   * @param {string|number} reconciliationId - Reconciliation ID
   * @param {string|number} itemId - Item ID
   * @param {Object} data - Updated item data
   * @returns {Promise<Object>} - Updated item
   */
  async updateReconciliationItem(reconciliationId, itemId, data) {
    try {
      const response = await api.put(
        `${RECONCILIATION_BASE_URL}/${reconciliationId}/items/${itemId}`,
        data
      );
      return response.data;
    } catch (error) {
      console.error(
        `Error updating item ${itemId} in reconciliation ${reconciliationId}:`,
        error
      );
      throw error;
    }
  },

  /**
   * Delete a reconciliation item
   * @param {string|number} reconciliationId - Reconciliation ID
   * @param {string|number} itemId - Item ID
   * @returns {Promise<void>}
   */
  async deleteReconciliationItem(reconciliationId, itemId) {
    try {
      await api.delete(
        `${RECONCILIATION_BASE_URL}/${reconciliationId}/items/${itemId}`
      );
    } catch (error) {
      console.error(
        `Error deleting item ${itemId} from reconciliation ${reconciliationId}:`,
        error
      );
      throw error;
    }
  },

  /**
   * Get reconciliation rules
   * @param {Object} params - Query parameters (accountId, isActive, etc.)
   * @returns {Promise<Array>} - List of reconciliation rules
   */
  async getReconciliationRules(params = {}) {
    try {
      const response = await api.get(`${RECONCILIATION_BASE_URL}/rules`, { params });
      return response.data;
    } catch (error) {
      console.error('Error fetching reconciliation rules:', error);
      throw error;
    }
  },

  /**
   * Create a reconciliation rule
   * @param {Object} data - Rule data
   * @returns {Promise<Object>} - Created rule
   */
  async createReconciliationRule(data) {
    try {
      const response = await api.post(`${RECONCILIATION_BASE_URL}/rules`, data);
      return response.data;
    } catch (error) {
      console.error('Error creating reconciliation rule:', error);
      throw error;
    }
  },

  /**
   * Update a reconciliation rule
   * @param {string|number} ruleId - Rule ID
   * @param {Object} data - Updated rule data
   * @returns {Promise<Object>} - Updated rule
   */
  async updateReconciliationRule(ruleId, data) {
    try {
      const response = await api.put(
        `${RECONCILIATION_BASE_URL}/rules/${ruleId}`,
        data
      );
      return response.data;
    } catch (error) {
      console.error(`Error updating reconciliation rule ${ruleId}:`, error);
      throw error;
    }
  },

  /**
   * Delete a reconciliation rule
   * @param {string|number} ruleId - Rule ID
   * @returns {Promise<void>}
   */
  async deleteReconciliationRule(ruleId) {
    try {
      await api.delete(`${RECONCILIATION_BASE_URL}/rules/${ruleId}`);
    } catch (error) {
      console.error(`Error deleting reconciliation rule ${ruleId}:`, error);
      throw error;
    }
  },

  /**
   * Get audit logs for a reconciliation
   * @param {string|number} reconciliationId - Reconciliation ID
   * @param {Object} params - Query parameters (action, startDate, endDate, etc.)
   * @returns {Promise<Array>} - List of audit logs
   */
  async getReconciliationAuditLogs(reconciliationId, params = {}) {
    try {
      const response = await api.get(
        `${RECONCILIATION_BASE_URL}/${reconciliationId}/audit-logs`,
        { params }
      );
      return response.data;
    } catch (error) {
      console.error(
        `Error fetching audit logs for reconciliation ${reconciliationId}:`,
        error
      );
      throw error;
    }
  },

  /**
   * Export reconciliation data
   * @param {string|number} reconciliationId - Reconciliation ID
   * @param {string} format - Export format (pdf, csv, xlsx)
   * @returns {Promise<Blob>} - Exported file data
   */
  async exportReconciliation(reconciliationId, format = 'pdf') {
    try {
      const response = await api.get(
        `${RECONCILIATION_BASE_URL}/${reconciliationId}/export`,
        {
          params: { format },
          responseType: 'blob'
        }
      );
      return response.data;
    } catch (error) {
      console.error(`Error exporting reconciliation ${reconciliationId}:`, error);
      throw error;
    }
  },

  /**
   * Auto-match reconciliation items
   * @param {string|number} reconciliationId - Reconciliation ID
   * @returns {Promise<Array>} - List of matched items
   */
  async autoMatchItems(reconciliationId) {
    try {
      const response = await api.post(
        `${RECONCILIATION_BASE_URL}/${reconciliationId}/auto-match`
      );
      return response.data;
    } catch (error) {
      console.error(
        `Error auto-matching items for reconciliation ${reconciliationId}:`,
        error
      );
      throw error;
    }
  },

  /**
   * Complete a reconciliation
   * @param {string|number} reconciliationId - Reconciliation ID
   * @returns {Promise<Object>} - Updated reconciliation
   */
  async completeReconciliation(reconciliationId) {
    try {
      const response = await api.post(
        `${RECONCILIATION_BASE_URL}/${reconciliationId}/complete`
      );
      return response.data;
    } catch (error) {
      console.error(`Error completing reconciliation ${reconciliationId}:`, error);
      throw error;
    }
  },

  /**
   * Reopen a completed reconciliation
   * @param {string|number} reconciliationId - Reconciliation ID
   * @returns {Promise<Object>} - Updated reconciliation
   */
  async reopenReconciliation(reconciliationId) {
    try {
      const response = await api.post(
        `${RECONCILIATION_BASE_URL}/${reconciliationId}/reopen`
      );
      return response.data;
    } catch (error) {
      console.error(`Error reopening reconciliation ${reconciliationId}:`, error);
      throw error;
    }
  },

  /**
   * Get reconciliation statistics
   * @param {Object} params - Query parameters (accountId, startDate, endDate, etc.)
   * @returns {Promise<Object>} - Statistics data
   */
  async getReconciliationStats(params = {}) {
    try {
      const response = await api.get(
        `${RECONCILIATION_BASE_URL}/stats`,
        { params }
      );
      return response.data;
    } catch (error) {
      console.error('Error fetching reconciliation statistics:', error);
      throw error;
    }
  },

  /**
   * Import reconciliation data from a file
   * @param {string|number} reconciliationId - Reconciliation ID
   * @param {File} file - File to import
   * @param {string} format - Import format (csv, xlsx)
   * @returns {Promise<Object>} - Import result
   */
  async importReconciliationData(reconciliationId, file, format = 'csv') {
    try {
      const formData = new FormData();
      formData.append('file', file);
      formData.append('format', format);

      const response = await api.post(
        `${RECONCILIATION_BASE_URL}/${reconciliationId}/import`,
        formData,
        {
          headers: {
            'Content-Type': 'multipart/form-data'
          }
        }
      );
      return response.data;
    } catch (error) {
      console.error(
        `Error importing data for reconciliation ${reconciliationId}:`,
        error
      );
      throw error;
    }
  }
};

File: backend/app/services/accounting/reconciliation_service.py
"""
Reconciliation Service

This module provides services for managing account reconciliations.
"""
from typing import Any, Dict, List, Optional, Type, TypeVar

from .. import models, schemas
from ..exceptions import (
from .account_balance_service import AccountBalanceService
from sqlalchemy import and_, or_, func, select
from sqlalchemy.orm import Session, joinedload
from uuid import UUID

from app.services.audit.audit_service import AuditService
from app.services.base import BaseService



    AccountNotFoundException,
    InvalidReconciliationStateException,
    JournalEntryNotFoundException,
    ReconciliationItemNotFoundException,
    ReconciliationNotFoundException,
)


class ReconciliationService(BaseService[models.Reconciliation, schemas.ReconciliationCreate, schemas.ReconciliationUpdate]):
    """Service for managing account reconciliations."""
    
    def __init__(self, db: Optional[Session] = None):
        """  Init  ."""
        """
        Initialize the service with an optional database session.
        
        Args:
            db: Optional database session. If not provided, a new one will be created.
        """
        super().__init__(models.Reconciliation)
        self.db = db
        self.audit_service = AuditService(db or self._get_db())
        self.balance_service = AccountBalanceService(db)
    
    def _get_db(self) -> Session:
        """ Get Db."""
        """Get a database session."""
        if self.db is None:
            from app.core.database import SessionLocal
            self.db = SessionLocal()
        return self.db
    
    def _get_reconciliation_or_404(self, reconciliation_id: UUID) -> models.Reconciliation:
        """ Get Reconciliation Or 404."""
        """Get a reconciliation by ID or raise appropriate exception."""
        return self._get_or_404(
            self._get_db(), 
            models.Reconciliation, 
            reconciliation_id,
            ReconciliationNotFoundException(reconciliation_id)
        )
    
    def _get_reconciliation_item_or_404(self, item_id: UUID) -> models.ReconciliationItem:
        """ Get Reconciliation Item Or 404."""
        """Get a reconciliation item by ID or raise appropriate exception."""
        return self._get_or_404(
            self._get_db(),
            models.ReconciliationItem,
            item_id,
            ReconciliationItemNotFoundException(item_id)
        )
    
    def _get_account_or_404(self, account_id: UUID) -> models.Account:
        """ Get Account Or 404."""
        """Get an account by ID or raise appropriate exception."""
        return self._get_or_404(
            self._get_db(),
            models.Account,
            account_id,
            AccountNotFoundException(account_id)
        )
        
    def _validate_reconciliation_state(
        """ Validate Reconciliation State."""
        self, 
        reconciliation: models.Reconciliation, 
        required_status: Optional[models.ReconciliationStatus] = None,
        check_items: bool = False
    ) -> None:
        """ Validate Reconciliation State."""
        """
        Validate reconciliation state.
        
        Args:
            reconciliation: The reconciliation to validate
            required_status: If provided, the reconciliation must be in this state
            check_items: If True, check for unmatched items when status is COMPLETED
            
        Raises:
            InvalidReconciliationStateException: If validation fails
        """
        if required_status and reconciliation.status != required_status:
            raise InvalidReconciliationStateException(
                f"Reconciliation must be in {required_status} state"
            )
            
        if check_items and reconciliation.status == models.ReconciliationStatus.COMPLETED:
            # Check for unmatched items only if reconciliation is being completed
            unmatched_items = self._count_unmatched_items(reconciliation.id)
            if unmatched_items > 0:
                raise InvalidReconciliationStateException(
                    f"Cannot complete reconciliation with {unmatched_items} unmatched items"
                )
    
    def _count_unmatched_items(self, reconciliation_id: UUID) -> int:
        """ Count Unmatched Items."""
        """
        Count unmatched items for a reconciliation.
        
        Args:
            reconciliation_id: ID of the reconciliation
            
        Returns:
            int: Number of unmatched items
        """
        db = self._get_db()
        return db.query(models.ReconciliationItem).filter(
            models.ReconciliationItem.reconciliation_id == reconciliation_id,
            models.ReconciliationItem.is_matched == False  # noqa: E712
        ).count()
        
    def _create_audit_log(
        """ Create Audit Log."""
        self,
        reconciliation_id: UUID,
        user_id: UUID,
        action: str,
        details: Dict[str, Any]
    ) -> None:
        """ Create Audit Log."""
        """
        Create an audit log entry for a reconciliation.
        
        Args:
            reconciliation_id: ID of the reconciliation
            user_id: ID of the user performing the action
            action: Action being performed
            details: Additional details about the action
            
        Returns:
            The created audit log entry
        """
        db = self._get_db()
        log = models.ReconciliationAuditLog(
            id=uuid4(),
            reconciliation_id=reconciliation_id,
            user_id=user_id,
            action=action,
            details=details
        )
        
        db.add(log)
        db.commit()
        db.refresh(log)
        return log
    def create_reconciliation(
        """Create Reconciliation."""
        self,
        reconciliation: schemas.ReconciliationCreate,
        user_id: UUID
    ) -> models.Reconciliation:
        """Create Reconciliation."""
        """
        Create a new reconciliation.
        
        Args:
            reconciliation: Reconciliation data to create
            user_id: ID of the user creating the reconciliation
            
        Returns:
            The created reconciliation
        """
        db = self._get_db()
        
        # Create the reconciliation record
        db_reconciliation = models.Reconciliation(
            **reconciliation.dict(exclude={"id", "created_by", "updated_by"}),
            id=uuid4(),
            status=models.ReconciliationStatus.DRAFT,
            created_by=user_id,
            updated_by=user_id
        )
        
        db.add(db_reconciliation)
        db.commit()
        db.refresh(db_reconciliation)
        
        # Log the creation using the audit service
        self.audit_service.log_action(
            action="reconciliation.create",
            resource_type="reconciliation",
            user_id=user_id,
            resource_id=str(db_reconciliation.id),
            new_values={"status": db_reconciliation.status.value}
        )
        
        return db_reconciliation
    
    def get_reconciliation(
        """Get Reconciliation."""
        self,
        reconciliation_id: UUID,
        user_id: UUID
    ) -> Optional[models.Reconciliation]:
        """Get Reconciliation."""
        """
        Get a reconciliation by ID with related data.
        
        Args:
            reconciliation_id: ID of the reconciliation to retrieve
            user_id: ID of the user making the request
            
        Returns:
            The reconciliation with account and items loaded, or None if not found
        """
        db = self._get_db()
        return db.query(models.Reconciliation).filter(
            models.Reconciliation.id == reconciliation_id
        ).options(
            joinedload(models.Reconciliation.account),
            joinedload(models.Reconciliation.items)
        ).first()
    
    def get_reconciliations(
        """Get Reconciliations."""
        self,
        user_id: UUID,
        account_id: Optional[UUID] = None,
        status: Optional[models.ReconciliationStatus] = None,
        start_date: Optional[datetime] = None,
        end_date: Optional[datetime] = None,
        skip: int = 0,
        limit: int = 100
    ) -> List[models.Reconciliation]:
        """Get Reconciliations."""
        """
        Get a list of reconciliations with optional filtering.
        
        Args:
            user_id: ID of the user making the request
            account_id: Optional account ID to filter by
            status: Optional status to filter by
            start_date: Optional start date to filter reconciliations
            end_date: Optional end date to filter reconciliations
            skip: Number of records to skip (for pagination)
            limit: Maximum number of records to return (for pagination)
            
        Returns:
            List of reconciliation records with account information
        """
        db = self._get_db()
        query = db.query(models.Reconciliation)
        
        # Apply filters
        if account_id:
            query = query.filter(models.Reconciliation.account_id == account_id)
            
        if status:
            query = query.filter(models.Reconciliation.status == status)
            
        if start_date:
            query = query.filter(models.Reconciliation.statement_date >= start_date)
            
        if end_date:
            # Include the entire end date
            end_date = end_date.replace(hour=23, minute=59, second=59)
            query = query.filter(models.Reconciliation.statement_date <= end_date)
            
        return query.options(
            joinedload(models.Reconciliation.account)
        ).order_by(
            models.Reconciliation.statement_date.desc()
        ).offset(skip).limit(limit).all()
    
    def update_reconciliation(
        """Update Reconciliation."""
        self,
        reconciliation_id: UUID,
        reconciliation_update: schemas.ReconciliationUpdate,
        user_id: UUID
    ) -> models.Reconciliation:
        """Update Reconciliation."""
        """
        Update a reconciliation.
        
        Args:
            reconciliation_id: ID of the reconciliation to update
            reconciliation_update: Data to update the reconciliation with
            user_id: ID of the user performing the update
            
        Returns:
            The updated reconciliation record
            
        Raises:
            HTTPException: If reconciliation is not found or not in a valid state for update
        """
        db = self._get_db()
        db_reconciliation = self._get_or_404(db, models.Reconciliation, reconciliation_id)
        
        # Store old values for audit log
        update_data = reconciliation_update.dict(exclude_unset=True)
        old_values = {
            k: getattr(db_reconciliation, k) 
            for k in update_data.keys()
            if hasattr(db_reconciliation, k)
        }
        
        # Validate reconciliation state before update
        self._validate_reconciliation_state(
            db_reconciliation,
            check_items=False
        )
        
        # Update the record
        db_reconciliation.updated_by = user_id
        db_reconciliation.updated_at = datetime.utcnow()
        
        for field, value in update_data.items():
            if hasattr(db_reconciliation, field):
                setattr(db_reconciliation, field, value)
        
        db.add(db_reconciliation)
        db.commit()
        db.refresh(db_reconciliation)
        
        # Log the update action
        self.audit_service.log_action(
            action="reconciliation.update",
            resource_type="reconciliation",
            user_id=user_id,
            resource_id=str(reconciliation_id),
            old_values=old_values,
            new_values=update_data
        )
        
        return db_reconciliation
    
    def delete_reconciliation(
        """Delete Reconciliation."""
        self,
        reconciliation_id: UUID,
        user_id: UUID
    ) -> bool:
        """Delete Reconciliation."""
        """
        Delete a reconciliation.
        
        Args:
            reconciliation_id: ID of the reconciliation to delete
            user_id: ID of the user performing the deletion
            
        Returns:
            bool: True if deletion was successful
            
        Raises:
            HTTPException: If reconciliation is not found or not in a valid state for deletion
        """
        db = self._get_db()
        db_reconciliation = self._get_or_404(db, models.Reconciliation, reconciliation_id)
        
        # Validate state before deletion
        self._validate_reconciliation_state(db_reconciliation)
        
        # Store data for audit log before deletion
        reconciliation_data = {
            'account_id': str(db_reconciliation.account_id),
            'statement_date': db_reconciliation.statement_date.isoformat(),
            'status': db_reconciliation.status.value
        }
        
        # Log the deletion action
        self.audit_service.log_action(
            action="reconciliation.delete",
            resource_type="reconciliation",
            user_id=user_id,
            resource_id=str(reconciliation_id),
            old_values=reconciliation_data
        )
        
        # Delete the reconciliation
        db.delete(db_reconciliation)
        db.commit()
        
        return True
    
    def add_reconciliation_item(
        """Add Reconciliation Item."""
        self,
        reconciliation_id: UUID,
        item: schemas.ReconciliationItemCreate,
        user_id: UUID
    ) -> models.ReconciliationItem:
        """Add Reconciliation Item."""
        """
        Add an item to a reconciliation.
        
        Args:
            reconciliation_id: ID of the reconciliation to add the item to
            item: Data for the new reconciliation item
            user_id: ID of the user adding the item
            
        Returns:
            The created reconciliation item
            
        Raises:
            HTTPException: If reconciliation is not found or not in a valid state for modification
        """
        db = self._get_db()
        
        # Get the reconciliation
        reconciliation = self._get_reconciliation_or_404(reconciliation_id)
        
        # Validate reconciliation state
        self._validate_reconciliation_state(reconciliation)
        
        # Create the item
        db_item = models.ReconciliationItem(
            **item.dict(),
            id=uuid4(),
            reconciliation_id=reconciliation_id,
            created_by=user_id,
            updated_by=user_id,
            created_at=datetime.utcnow(),
            updated_at=datetime.utcnow()
        )
        
        db.add(db_item)
        db.commit()
        db.refresh(db_item)
        
        # Log the item addition
        self.audit_service.log_action(
            action="reconciliation.item.add",
            resource_type="reconciliation_item",
            user_id=user_id,
            resource_id=str(db_item.id),
            parent_resource_type="reconciliation",
            parent_resource_id=str(reconciliation_id),
            new_values={
                "item_type": db_item.item_type.value,
                "amount": float(db_item.amount),
                "transaction_date": db_item.transaction_date.isoformat() if db_item.transaction_date else None,
                "description": db_item.description
            }
        )
        
        return db_item
    
    def update_reconciliation_item(
        """Update Reconciliation Item."""
        self,
        item_id: UUID,
        item_update: schemas.ReconciliationItemUpdate,
        user_id: UUID
    ) -> models.ReconciliationItem:
        """Update Reconciliation Item."""
        """
        Update a reconciliation item.
        
        Args:
            item_id: ID of the item to update
            item_update: Data to update the item with
            user_id: ID of the user performing the update
            
        Returns:
            The updated reconciliation item
            
        Raises:
            HTTPException: If item or its reconciliation is not found or not in a valid state for modification
        """
        db = self._get_db()
        
        # Get the item and its reconciliation
        db_item = self._get_reconciliation_item_or_404(item_id)
        reconciliation = self._get_reconciliation_or_404(db_item.reconciliation_id)
        
        # Validate reconciliation state
        self._validate_reconciliation_state(reconciliation)
        
        # Store old values for audit log
        update_data = item_update.dict(exclude_unset=True)
        old_values = {
            k: getattr(db_item, k) 
            for k in update_data.keys()
            if hasattr(db_item, k)
        }
        
        # Update the item
        for field, value in update_data.items():
            if hasattr(db_item, field):
                setattr(db_item, field, value)
        
        db_item.updated_by = user_id
        db_item.updated_at = datetime.utcnow()
        
        db.add(db_item)
        db.commit()
        db.refresh(db_item)
        
        # Log the item update
        self.audit_service.log_action(
            action="reconciliation.item.update",
            resource_type="reconciliation_item",
            user_id=user_id,
            resource_id=str(item_id),
            parent_resource_type="reconciliation",
            parent_resource_id=str(db_item.reconciliation_id),
            old_values=old_values,
            new_values={
                k: v.isoformat() if hasattr(v, 'isoformat') and callable(v.isoformat) 
                else v for k, v in update_data.items()
            }
        )
        
        # Delete the item
        db.delete(db_item)
        db.commit()

    def complete_reconciliation(
        """Complete Reconciliation."""
        self,
        reconciliation_id: UUID,
        user_id: UUID
    ) -> models.Reconciliation:
        """Complete Reconciliation."""
        """
        Mark a reconciliation as completed.
        
        Args:
            reconciliation_id: ID of the reconciliation to complete
            user_id: ID of the user completing the reconciliation
            
        Returns:
            The completed reconciliation record
            
        Raises:
            HTTPException: If reconciliation is not found, not in DRAFT status, or has unmatched items
        """
        db = self._get_db()
        db_reconciliation = self._get_or_404(db, models.Reconciliation, reconciliation_id)
        
        # Validate reconciliation can be completed
        self._validate_reconciliation_state(
            db_reconciliation,
            required_status=models.ReconciliationStatus.DRAFT,
            check_items=True  # This will check for unmatched items
        )
        
        # Store old values for audit log
        old_status = db_reconciliation.status
        
        # Update reconciliation status
        db_reconciliation.status = models.ReconciliationStatus.COMPLETED
        db_reconciliation.completed_by = user_id
        db_reconciliation.completed_at = datetime.utcnow()
        db_reconciliation.updated_by = user_id
        db_reconciliation.updated_at = datetime.utcnow()
        
        db.add(db_reconciliation)
        db.commit()
        db.refresh(db_reconciliation)
        
        # Log the completion
        self.audit_service.log_action(
            action="reconciliation.complete",
            resource_type="reconciliation",
            user_id=user_id,
            resource_id=str(reconciliation_id),
            old_values={"status": old_status.value},
            new_values={
                "status": db_reconciliation.status.value,
                "completed_at": db_reconciliation.completed_at.isoformat(),
                "completed_by": str(user_id)
            }
        )
        
        return db_reconciliation
    
    def get_unreconciled_transactions(
        """Get Unreconciled Transactions."""
        self,
        account_id: UUID,
        start_date: datetime,
        end_date: datetime,
        user_id: UUID
    ) -> List[Dict[str, Any]]:
        """Get Unreconciled Transactions."""
        """
        Get unreconciled transactions for an account within a date range.
        
        Args:
            account_id: ID of the account to get transactions for
            start_date: Start of the date range (inclusive)
            end_date: End of the date range (inclusive)
            user_id: ID of the user making the request (for audit purposes)
            
        Returns:
            List of dictionaries containing transaction details
            
        Raises:
            HTTPException: If account is not found
        """
        db = self._get_db()
        
        # Verify account exists
        self._get_account_or_404(account_id)
        
        # Log the query for audit purposes
        self.audit_service.log_action(
            action="reconciliation.transactions.query",
            resource_type="account",
            resource_id=str(account_id),
            user_id=user_id,
            details={
                "start_date": start_date.isoformat(),
                "end_date": end_date.isoformat(),
                "status": "unreconciled"
            }
        )
        
        # Get all transactions for the account in the date range
        transactions = db.query(models.Transaction).filter(
            models.Transaction.account_id == account_id,
            models.Transaction.transaction_date.between(start_date, end_date),
            models.Transaction.reconciled == False  # noqa: E712
        ).order_by(
            models.Transaction.transaction_date.asc()
        ).all()
        
        return [
            {
                "id": str(txn.id),
                "transaction_date": txn.transaction_date.isoformat(),
                "amount": float(txn.amount),
                "description": txn.description,
                "reference": txn.reference,
                "transaction_type": txn.transaction_type.value,
                "created_at": txn.created_at.isoformat() if txn.created_at else None,
                "updated_at": txn.updated_at.isoformat() if txn.updated_at else None
            }
            for txn in transactions
        ]


# Create a singleton instance
reconciliation_service = ReconciliationService()
